### PR TITLE
generate json output

### PR DESCRIPTION
**Pull request by sigu**
Generate json sbom files by adding `.json` to the output filename

## Note:
Merging/appending of json files is still not possible using the nodejs module https://github.com/CycloneDX/cyclonedx-node-module/issues/90

### CODE REVIEW

File: test/sbom/cyclonedx_test.exs [hunk 1]
```diff
@@ -85,5 +85,96 @@ defmodule SBoM.CycloneDXTest do
...
+            version: "0.0.1",
+            description: "support for CycloneDX in json support",
+            purl: "pkg:hex/name@0.0.1",
+            licenses: ["Apache-2.0"]
+          }
+        ]
+        |> bom(format: :json)
+        |> to_string()
+
+      assert json =~ "{\n  \"bomFormat\": \"CycloneDX\""
```
Comment (voltone) [inline test/sbom/cyclonedx_test.exs]: It seems a bit brittle to assert on a JSON literal with whitespace. I would prefer asserting that the result can be decoded by Jason, and then assert on values in the resulting map:

```elixir
  assert {:ok, decoded} = Jason.parse(json)
  assert decoded["bomFormat"] == "CycloneDX"
```

Same thing for the other JSON assertions in this file.

File: lib/mix/tasks/sbom.cyclonedx.ex [hunk 2]
```diff
@@ -91,6 +105,20 @@ defmodule Mix.Tasks.Sbom.Cyclonedx do
...
+    else
+      opts
+    end
+  end
+
+  defp validate_json_schema(opts) do
+    if opts[:format] == :json and opts[:schema] == "1.1" do
+      shell = Mix.shell()
+
+      shell.error("Json is not supported in version 1.1")
```
Comment (voltone) [inline lib/mix/tasks/sbom.cyclonedx.ex]:
```suggestion
      shell.error("JSON is not supported in version 1.1")
```

File: lib/mix/tasks/sbom.cyclonedx.ex [hunk 3]
```diff
@@ -91,6 +105,20 @@ defmodule Mix.Tasks.Sbom.Cyclonedx do
...
+
+      Mix.raise("JSON is NOT supported for version 1.1 of cyclonedx.")
```
Comment (voltone) [inline lib/mix/tasks/sbom.cyclonedx.ex]:
```suggestion
      Mix.raise("JSON is NOT supported for version 1.1 of CycloneDX.")
```

File: lib/sbom/cyclonedx.ex [hunk 4]
```diff
@@ -1,94 +1,34 @@
 defmodule SBoM.CycloneDX do
   @moduledoc """
-  Generate a CycloneDX SBoM in XML format.
+  Generate a CycloneDX SBoM in XML and JSON format.
```
Comment (voltone) [inline lib/sbom/cyclonedx.ex]:
```suggestion
  Generate a CycloneDX SBoM in XML or JSON format.
```

**Review by voltone — CHANGES_REQUESTED** [review]
Looking good, thanks! A few tiny change requests, and a suggestion for more robust testing, below...

**Comment by sigu** [comment]
@voltone , these comments have been worked on, thanks to @TraceyOnim

**Comment by voltone** [comment]
Thanks for the updates. I was going to test this with OWASP DependencyTrack 4.0 beta, but I ran into a few issues.

One is that the JSON encoder assumes all fields are present, but some component fields are not always available, so something like `component.description` might fail. That's easy enough to fix.

But more fundamentally, I prefer to install the SBoM Mix task globally, as an archive (`mix archive.install hex sbom`), rather than adding it as a project dependency. Unfortunately, such Mix tasks cannot rely on dependencies, so the Jason package is unavailable. Need to think about this: either we'd have to give up on using this as an archive (at least when JSON output is desired), or we add just enough JSON builder support, like we do for XML...

**Comment by sigu** [comment]
I am closing this since we can use https://github.com/CycloneDX/cyclonedx-cli to convert to different formats

**Pull request closed without merging**